server/src/Postgrid/Cache.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Postgrid.Cache
+    ( CachedApiEndpoint(..)
+    , QueryCache(..)
+    , cached
+    , newPostgridQueryCache
+    ) where
+
+import Data.Aeson (encode)
+import Data.ByteString.Lazy (ByteString)
+import Data.IORef (atomicModifyIORef')
+import Data.Hashable (hash)
+import qualified Data.LruCache as Lru (lookup, insert)
+import Data.LruCache.IO (LruHandle(..), StripedLruHandle(..), newStripedLruHandle)
+import qualified Data.Vector as Vector
+
+import Postgrid.API (PostgridError)
+import Postgrid.API.Types
+
+data CachedApiEndpoint req resp where
+    CachedVerifyStructuredAddress :: CachedApiEndpoint VerifyStructuredAddressRequest VerifyStructuredAddressResponse
+
+cacheKey :: CachedApiEndpoint req resp -> req -> ByteString
+cacheKey CachedVerifyStructuredAddress = encode
+
+data CacheEntry where
+    CacheEntry :: CachedApiEndpoint req resp -> resp -> CacheEntry
+
+-- | In-memory cache for Postgrid API query results. It stores either a successful encoded response or an error.
+-- Encoded response is stored for the sake of memory efficiency.
+newtype QueryCache = QueryCache { unQueryCache :: StripedLruHandle ByteString CacheEntry }
+
+newPostgridQueryCache :: Int -> Int -> IO QueryCache
+newPostgridQueryCache numStripes size = QueryCache <$> newStripedLruHandle numStripes size
+
+-- | A helper for 'QueryCache' that checks the cache for a given request
+-- and either returns the cached response or runs the provided action to fetch it.
+-- Errors are not cached.
+cached :: QueryCache -> req -> CachedApiEndpoint req resp -> IO (Either PostgridError resp) -> IO (Either PostgridError resp)
```

Comment:
@rinn7e , I've realized that Postgrid query cache was caching both successful responses and errors, while it is a bad idea to cache errors because they can be transient 😅

May I ask you to take a look at this module once again?

Reply:
Sure, looks fine to me. 1 question though, Lru is in-memory cache only right?